xds/src/main/java/com/linecorp/armeria/xds/XdsEndpointGroup.java

```diff
@@ -66,13 +83,57 @@ public static EndpointGroup of(ListenerRoot listenerRoot) {
             if (routeSnapshot == null) {
                 return;
             }
-            final List<ClusterLoadAssignment> endpoints =
-                    routeSnapshot.clusterSnapshots().stream()
-                                 .map(ClusterSnapshot::endpointSnapshot)
-                                 .filter(Objects::nonNull)
-                                 .map(endpointSnapshot -> endpointSnapshot.xdsResource().resource())
-                                 .collect(Collectors.toList());
-            setEndpoints(convertEndpoints(endpoints));
+
+            final List<ClusterSnapshot> clusterSnapshots = routeSnapshot.clusterSnapshots();
+            if (clusterSnapshots.isEmpty()) {
+                return;
+            }
+
+            if (clusterSnapshots.size() > 1) {
+                // Currently, the first cluster is only used until we implement EndpointGroupSelector.
+                logger.debug("The clusters from the second one are ignored. Ignored clusters: {}",
+                             clusterSnapshots.subList(1, clusterSnapshots.size()));
+            }
+
+            final ClusterSnapshot clusterSnapshot = clusterSnapshots.get(0);
+            final EndpointSnapshot endpointSnapshot = clusterSnapshot.endpointSnapshot();
+            if (endpointSnapshot == null) {
+                logger.debug("Skipping cluster without an endpoint. {}", clusterSnapshot.xdsResource());
+                return;
+            }
+
+            final Struct filterMetadata = filterMetadata(clusterSnapshot);
+            if (filterMetadata.getFieldsCount() == 0) {
+                // No metadata. Use the whole endpoints.
+                setEndpoints(endpointSnapshot);
+                return;
+            }
+
+            final Cluster cluster = clusterSnapshot.xdsResource().resource();
+            final LbSubsetConfig lbSubsetConfig = cluster.getLbSubsetConfig();
+            if (lbSubsetConfig == LbSubsetConfig.getDefaultInstance()) {
+                // No lbSubsetConfig. Use the whole endpoints.
+                setEndpoints(endpointSnapshot);
+                return;
+            }
+            final LbSubsetFallbackPolicy fallbackPolicy = lbSubsetConfig.getFallbackPolicy();
+            if (fallbackPolicy != LbSubsetFallbackPolicy.ANY_ENDPOINT) {
```

Comment:
Note: I guess no need to handle this in this PR, but this setting can technically be overridden by `LbSubsetSelector#getFallbackPolicy` which can make this a false-positive warning.

Reply:
Yeah, let's handle it in another PR. I didn't implement the feature fully. 😉